What's an UUID article (#461)
unicorn-utterances/unicorn-utterances: content/blog/what-are-uuids/index.md

```
a5abec44-7ce0-437c-972a-cf451b4fde2b
```

Two of these characters include information about which kind of UUID it is:

```
xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx
```

Here, `M` is the UUID version number, while `N` is the variant of the UUID version. Think of this as information encoded within the UUID _about_ the UUID generation process that yielded the output value. 

> While we'll explain what the UUID "version" is in the next section, let's quickly explain what the "variant" means:
>
> - If the variant is `0` through `7`, the UUID is backward compatible with [**very** old computer systems from the 1980s](https://en.wikipedia.org/wiki/Apollo_Computer).
> - If the variant is `8` through `b`, it means the UUID is part of [the "RFC 4122" standard](https://www.ietf.org/rfc/rfc4122.txt)
> - If the variant is `c` or `d`, it means the UUID is compatible with early Windows systems
> - Variants `e` and `f` are reserved for future UUID versions

This means that we can take the previous UUID:

```
a5abec44-7ce0-437c-972a-cf451b4fde2b
```

And determine that this is a UUID with the version of `4` and the variant of `9`.

> > Speaking of "UUID version numbers", what are those? I know we outlined earlier that there were different ways of generating a UUID; what are they?

Author: Strongly disagree. It's a quote from the perspective of the reader.

Let's say you're creating a database of URLs tracked on your website. You want to use the URL as [the primary key in your database](https://en.wikipedia.org/wiki/Primary_key) but don't want to use the full URL, as it will bloat your key's storage requirements.

On the other side of the coin: You can't use any random data to store as your URL's primary key since it could introduce duplicate URL entries into your database table. 

Is there a good way to generate a unique ID based on a string that will output the same ID if the inputs are the same?

Intro: UUIDv3 and UUIDv5.

Both of these UUID versions take the following information:

- A pre-defined UUID "namespace"
- Any string for a "name"

> The UUID specification establishes 4 pre-defined namespaces for common use cases. The pre-defined namespaces are:
>
> - [DNS](https://en.wikipedia.org/wiki/Domain_Name_System): `6ba7b810-9dad-11d1-80b4-00c04fd430c8`
> - [URL](https://en.wikipedia.org/wiki/URL): `6ba7b811-9dad-11d1-80b4-00c04fd430c8`
> - [OID](https://en.wikipedia.org/wiki/Object_identifier): `6ba7b812-9dad-11d1-80b4-00c04fd430c8`
> - [X.500 DN](https://en.wikipedia.org/wiki/X.500): `6ba7b814-9dad-11d1-80b4-00c04fd430c8`

> 

Reviewer: Could be useful to mention that one can create a namespace UUID if they want, and that these 4 are not the only ones.

Author: ```suggestion
>
> You may also use your own hardcoded UUID as a namespace
```

These UUID versions output a UUID that contains a hash of the namespace and name concatenated together.

You can think of the generation algorithm for both of these UUID versions as the following:

```javascript
UUID = hash(NAMESPACE + NAME)
```

> 

Reviewer: **Additional explanation:**
The reason a namespace is needed (instead of just hashing the name alone) is that even if the name is unique within a certain set of items (or "namespace" of items) it's likely that two different items from two different namespaces share the same name.
So to avoid duplicate UUIDs across ALL namespaces, we add the namespace UUID and hash the combination.
_And that allows us to generate the same UUID with the same input each time, which means we don't even have to store the value, we can generate them on the go._

For example, here's UUIDv3, which uses [MD5](https://en.wikipedia.org/wiki/MD5) to hash the concatenated value:

![A UUID broken down into "MD5 High", a dash, "MD5 High", a dash, "Version", "MD5 Mid", a dash, "Variant", "MD5 Low", a dash, and finally another "MD5 Low". An example UUIDv3 might be "a6a09ffa-9d61-3a0a-84a8-b27dd6dcf32f"](./UUIDv3.svg)

Compare this to UUIDv5, which uses [SHA-1](https://en.wikipedia.org/wiki/SHA-1) to generate the hash:

![// TODO: Write](./UUIDv5.svg)

### When should you use UUIDv3 vs. UUIDv5?

When provided the option, when should you use UUIDv3 against UUIDv5? When should you do the opposite?

**You should use UUIDv3 when**:

- Performance is key, or system resources are limited
- It's not critical that `name` or `namespace` contents can be decoded

This is because UUIDv3 uses the less secure MD5 hashing method for its contents. 

Likewise, **you should use UUIDv5 when**:

- It's critical that `name` or `namespace` contents cannot be decoded
- Performance is less of a concern

As UUIDv5's SHA-1 is significantly more secure with hashing its contents.

### Pros of UUIDv3 and UUIDv5

> - **Contents are stable**: In instances where you need the input to match the output every time, there are no other UUIDs that can fulfill this use case.

Author: ```suggestion
- **Contents are stable**: In instances where you need the output to match the input every time, no other UUID versions can fulfill this use case.
```
